gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/Client.java

```diff
@@ -742,19 +750,42 @@ protected Connection chooseConnection(final RequestMessage msg) throws TimeoutEx
          */
         @Override
         protected void initializeImplementation() {
```

Comment:
So we are changing the behaviour of this function?

If I understand correctly:
This function used to grab a random 'available' host, then initialize the connection for that host.
This function now attempts to grab a random 'available' host, then initialize the connection for that host, if there isn't a host available, it instead grabs a random 'unavailable' host, then initializes the connection for that host (presumably this means it is actually 'available'.

Is that right?

Reply:
Nit: Could do something like:
```
            // chooses an available host at random
            if (cluster.allHosts().isEmpty()) {
                // throw exception.
            }

            final List<Host> hosts = new ArrayList<>(cluster.allHosts());
            Collections.shuffle(hosts);
            Optional<Host> host = hosts.stream().filter(Host::isAvailable).findFirst();
            
            final CompletableFuture<Void> completableFuture =
                    CompletableFuture.runAsync(() -> initializeConnectionSetupForHost.accept(host.orElse(hosts.get(0))), cluster.executor()); // note **host.orElse(hosts.get(0))**
```
This is my personal preference, I think it reads nicer. But up to you.

Reply:
Yes, changing it so if there isn't any "available" hosts, it will look into all hosts. Technically if you started with a session client, no host will be marked available, and I think the only case there is available hosts is if a cluster client was started before it that marks some hosts as available. 
I think your change reads nicer as well.